ddl/schema.go

```diff
@@ -70,6 +70,48 @@ func onCreateSchema(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 	}
 }
 
+func rollbackDropSchema(t *meta.Meta, job *model.Job, dbInfo *model.DBInfo) (ver int64, err error) {
+	// If dbInfo schema state is public, no need to rollback.
+	if dbInfo.State == model.StatePublic {
+		job.State = model.JobStateRollbackDone
+		job.SchemaState = dbInfo.State
+		return ver, nil
+	}
+	// Recover database.
+	// Change type to ActionCreateSchema for infoschema to load schema diff.
+	job.Type = model.ActionCreateSchema
+	defer func() {
+		job.Type = model.ActionDropSchema
+	}()
+	dbInfo.State = model.StatePublic
+	err = t.UpdateDatabase(dbInfo)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	ver, err = updateSchemaVersion(t, job)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	// Recover table.
+	var tables []*model.TableInfo
+	tables, err = t.ListTables(job.SchemaID)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	// Change type to ActionCreateTable for infoschema to load schema diff.
+	job.Type = model.ActionCreateTable
+	for _, tblInfo := range tables {
+		job.TableID = tblInfo.ID
+		ver, err = updateSchemaVersion(t, job)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+	}
+	job.FinishDBJob(model.JobStateRollbackDone, model.StatePublic, ver, dbInfo)
```

Comment:
Actually, I have confusion here, `FinishDBJob` will add information to `job.BinlogInfo`, Is this use to write binlog? If so, here I only write dbInfo to binlog, but don't write table info  to binlog that table in this DB.
@winkyao @zimulala

Reply:
No tableInfo is ok for drop database.

Reply:
For insurance purposes, we need to discuss this with yulong.

Reply:
I think they will check the job's state. If the job's state is `JobStateRollbackDone`, it need not add binlog.